numpy/_core/src/common/npy_import.h

```diff
@@ -16,17 +79,22 @@
  * @param attr module attribute to cache.
  * @param cache Storage location for imported function.
  */
-static inline void
-npy_cache_import(const char *module, const char *attr, PyObject **cache)
-{
-    if (NPY_UNLIKELY(*cache == NULL)) {
-        PyObject *mod = PyImport_ImportModule(module);
-
-        if (mod != NULL) {
-            *cache = PyObject_GetAttrString(mod, attr);
-            Py_DECREF(mod);
+static inline int
+npy_cache_import_runtime(const char *module, const char *attr, PyObject **obj) {
+    if (!npy_atomic_load_ptr(obj)) {
+        PyThread_acquire_lock(npy_runtime_imports.import_mutex, WAIT_LOCK);
+        if (!npy_atomic_load_ptr(obj)) {
+            npy_atomic_store_ptr(obj, npy_import(module, attr));
+            if (obj == NULL) {
+                return -1;
+            }
         }
+        PyThread_release_lock(npy_runtime_imports.import_mutex);
```

Comment:
Good catch! Also, the `obj == NULL` check doesn't make sense -- it would need to be `*obj == NULL` to check the result.

But I think it would be better not to hold the lock around the `npy_import`:

* Imports in general can run arbitrary code, which may reentrantly trigger  `npy_cache_import_runtime`.
* Imports (i.e., `PyImport_ImportModule`) are already thread-safe internally

So I'd suggest writing this as:

```c
    if (!npy_atomic_load_ptr(obj)) {
        PyObject *value = npy_import(module, attr);
        if (value == NULL) {
            return -1;
        }
        PyThread_acquire_lock(npy_runtime_imports.import_mutex, WAIT_LOCK);
        if (!npy_atomic_load_ptr(obj)) {
            npy_atomic_store_ptr(obj, Py_NewRef(value));
        }
        PyThread_release_lock(npy_runtime_imports.import_mutex);
        Py_DECREF(value);
    }
```

Or you can get rid of the lock if you implement compare-exchange:

```c
if (!npy_atomic_load_ptr(obj)) {
        PyObject *value = npy_import(module, attr);
        if (value == NULL) {
            return -1;
        }
        PyObject *exepected = NULL;
        if (!npy_atomic_compare_exchange_ptr(obj, &expected, value)) {
            Py_DECREF(value);
        }
    }
```

Reply:
Thanks! I kept the lock since the next iteration will use PyMutex. Thankfully deadsnakes was updated over the weekend so we can test against a version with a public PyMutex now.